[css-transitions] `transition-property: all` does not apply to custom properties

graouts opened the pull request:
#### fcdce8ff40bbdf656f8d6d9247e9c94538a9dc2e
<pre>
[css-transitions] `transition-property: all` does not apply to custom properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=252312">https://bugs.webkit.org/show_bug.cgi?id=252312</a>

Reviewed by Antti Koivisto.

When setting `transition-property: all`, which also happens to be that property&apos;s initial value,
we would correctly consider all properties known within WebCore as a CSSPropertyID, but we would
fail to consider registered custom properties that could otherwise be interpolated.

We now iterate through all custom properties in both the before-change and after-change styles
to gather custom properties that can be interpolated (ie. either a SyntaxValue or SyntaxValueList)
and run the transition update logic on those.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-transition-property-all-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-transition-property-all.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/resources/utils.js:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSTransitions const):

Canonical link: <a href="https://commits.webkit.org/260384@main">https://commits.webkit.org/260384@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/faaa468e762d61bc379ab90f19f5f53c5b4b4894

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17247 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/41034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117291 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112053 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18595 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8535 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100374 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113934 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/41034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/95933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/41034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/10115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/41034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10838 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/16264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/41034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12420 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3905 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/Styleable.cpp:
```diff
@@ -672,6 +673,24 @@ void Styleable::updateCSSTransitions(const RenderStyle& currentStyle, const Rend
+        gatherAnimatableCustomPropertiesInMap(newStyle.inheritedCustomProperties());
+        gatherAnimatableCustomPropertiesInMap(newStyle.nonInheritedCustomProperties());
+        for (auto& customProperty : animatableCustomProperties)
+            updateCSSTransitionsForStyleableAndProperty(*this, customProperty, currentStyle, newStyle, generationTime);
```
you should test with `RenderStyle::customPropertiesEqual` before doing all this work.

graouts replied:
Rules for whether a property will yield to a transition update are complex and encapsulated in `updateCSSTransitionsForStyleableAndProperty()`. Even though two custom properties could have the same value, some transition properties could yield an update, for instance if the transition-duration changes in flight. I don't think we want to bring up any of the complicity of `updateCSSTransitionsForStyleableAndProperty()` up to here.

graouts committed:
[css-transitions] `transition-property: all` does not apply to custom properties
https://bugs.webkit.org/show_bug.cgi?id=252312

Reviewed by Antti Koivisto.

When setting `transition-property: all`, which also happens to be that property's initial value,
we would correctly consider all properties known within WebCore as a CSSPropertyID, but we would
fail to consider registered custom properties that could otherwise be interpolated.

We now iterate through all custom properties in both the before-change and after-change styles
to gather custom properties that can be interpolated (ie. either a SyntaxValue or SyntaxValueList)
and run the transition update logic on those.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-transition-property-all-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-transition-property-all.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/resources/utils.js:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSTransitions const):

Canonical link: https://commits.webkit.org/260384@main
